HW-05

alxiw opened the pull request:
Пятое домашнее задание по теме Selenium API

alxiw committed:
Add HW-05 and update AbstractPage and tests in HW-04

saint88 commented on homework-04/src/test/java/ru/mail/track/ui11/hw04/NameSecretHoroMailPageTest.java:
```diff
@@ -27,6 +27,7 @@ public void killSession() {
     public void check_the_ability_to_find_name(){
         new NameSecretHoroMailPage(driver)
                 .open()
+                .pageShallBeOpened()
```
Этот метод не нужно вызывать после опена, так как он должен вызываться в самом open()

alxiw replied:
В методе данная проверка присутствует, но забыл про неё, поэтому написал этот дублирующий метод 🙂
В общем, он не нужен, убрал его

saint88 commented on homework-04/src/test/java/ru/mail/track/ui11/hw04/NameSecretHoroMailPageTest.java:
```diff
@@ -37,6 +38,7 @@ public void check_the_ability_to_find_name(){
     public void check_the_ability_to_find_name_with_popup_suggestions(){
         new NameSecretHoroMailPage(driver)
                 .open()
+                .pageShallBeOpened()
```
Аналогично выше

alxiw replied:
Написал выше

saint88 commented on homework-04/src/test/java/ru/mail/track/ui11/hw04/NameSecretHoroMailPageTest.java:
```diff
@@ -49,6 +51,7 @@ public void check_the_ability_to_find_name_with_popup_suggestions(){
     public void check_the_ability_to_find_name_and_select_gender(){
         new NameSecretHoroMailPage(driver)
                 .open()
+                .pageShallBeOpened()
```
Аналогично выше

alxiw replied:
Написал выше

saint88 commented on homework-04/src/test/java/ru/mail/track/ui11/hw04/NameSecretHoroMailPageTest.java:
```diff
@@ -62,6 +65,7 @@ public void check_the_ability_to_find_names_by_letter_button(){
         String letter = "М";
         new NameSecretHoroMailPage(driver)
                 .open()
+                .pageShallBeOpened()
```
Аналогично выше

alxiw replied:
Написал выше

saint88 commented on homework-04/src/test/java/ru/mail/track/ui11/hw04/NewsPetsMailPageTest.java:
```diff
@@ -27,6 +27,7 @@ public void killSession() {
     public void check_show_more_button(){
         new NewsPetsMailPage(driver)
                 .open()
+                .pageShallBeOpened()
```
Аналогично выше

alxiw replied:
Написал выше

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -0,0 +1,107 @@
+import static org.junit.Assert.assertTrue;
+
+@UrlPattern("/prognoz/[\\w]{1,20}/")
+@PageUrl("/")
```
Почему у тебя PageUrl("/"). Он по такому урлу должен открыть главную: домен + "/". А у тебя шаблон на валидаторе страницы /prognoz/[\\w]{1,20}. Т.е. получается что-то не сходится.

alxiw replied:
Дело в том, что мы переходим по URL "/", но система сразу-же перебрасывает на страницу дефолтного города "/prognoz/<city_name>/", после чего срабатывает механизм проверки URL открытой страницы по шаблону регулярного выражения

saint88 replied:
Хорошо. Тогда отлично и оставляем так

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -0,0 +1,107 @@
+    @FindBy(css = ".pm-toolbar__search__input")
+    private WebElement searchInputField;
+
+    @FindBy(css = "span[class=\"pm-toolbar__suggests__group__item  pm-toolbar__suggests__group__item_with-caption\"]:first-child > span > span > span:first-child")
```
можно вместо двойных кавычек в локаторе, использовать одинарные кавычки, чтобы в стринге не экранировать их.

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -0,0 +1,107 @@
+    @FindBy(css = "span[class=\"pm-toolbar__suggests__group__item  pm-toolbar__suggests__group__item_with-caption\"]:first-child > span > span > span:first-child")
+    private WebElement cityElement;
+
+    @FindBy(xpath = "//div[@class='information__header__left__place']/h1[text()='Киев']")
```
А зачем тут харкодить город Киев. Его нужно передавать как параметр в локатор, потому что прийдет кейс на тестирование города внутри страны и все, будем писать новый локатор и метод.

alxiw replied:
Да, согласен 😊 
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -0,0 +1,107 @@
+    @FindBy(xpath = "//div[@class='information__header__left__place']/h1[text()='Киев']")
+    private WebElement cityOnPage;
+
+    @FindBy(css = "span[class=\"icon icon_fav-flag js-toggle_fav\"]")
```
Тут тоже лучше использовать одинарные кавычки, чтобы не экранировать. И такой селектор не пойдет. Вот пересоберут верстку и поменяют местами классы и все, локатор работать не будет. В общем его надо переписать.
Главные условия к локаторам: находит только нужные элементы на странице (уникальность) и адаптируемый к мелким изменениям на верстке, например смена порядка следования классов в элементе.

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -0,0 +1,107 @@
+    private WebElement addToFavoritesElement;
+
+    private final String cityDropdownSelector = "[data-class='pm-dropdown']";
+    private final String cityInFavorites = "//span[@class='city-select__item clearfix city-select__item_current']//span[@class='city-select__item__city']";
```
Аналогично выше. Локатор такой не пойдет

alxiw replied:
В чём проблема такого локатора? Мы же находимся на странице города, который только-что выбрали и добавили в "Избранное", и теперь ищем его в выпадающем списке по наличию в названии класса "city-select__item_current

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+import java.util.List;
+
+@UrlPattern("/prognoz/easter_island/[\\d]{1,2}-[\\w]{1,10}/#[\\d]{4}")
+@PageUrl("/prognoz/easter_island/10-june/#2015")
```
Зачем харкод? Вот тут нужно дату передавать в урл. Дату передавать нужно через LocalDate

alxiw replied:
ОК, исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+    @FindBy(xpath = "//span[contains(@class, 'calendar__control_prev')]")
+    private WebElement prevMonth;
+
+    @FindBy(xpath = "//span[contains(@class, 'calendar__control_next')]")
```
Точно такой же элемент написан выше, только в локаторе prev вместо next. Нужно сделать один общий локатор для кнопок навигации и передавать туда NavigationData. В enum определить что-то типа такого:
Next(">"),
Prev("<")

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+    @FindBy(xpath = "//span[contains(@class, 'calendar__control_next')]")
+    private WebElement nextMonth;
+
+    @FindBy(xpath = "//div[contains(@class, 'heading heading_minor heading_line')]")
```
Локатор переписать. Написал почему выше

alxiw replied:
Переписал

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+        return this;
+    }
+
+    public CalendarPogodaMailPage goToNextMonth() {
```
Такой метод уже есть выше. Этот надо убрать и сделать единый метод с передачей NavigationData

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    public CalendarPogodaMailPage goToDay(int day) {
+        driver.findElement(By.xpath("//table[contains(@class, 'calendar')]//a[text() = '" + day + "']")).click();
```
Такой локатор надо хранить в виде шаблона для String.format и как поле класса private final String

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+        return this;
+    }
+
+    public CalendarPogodaMailPage checkDayInList(String month, int day) {
```
Для работы с датами есть LocalDate, ее и надо передавать сюда

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -0,0 +1,76 @@
+
+    public CalendarPogodaMailPage checkDayInList(String month, int day) {
+        String firstThreeLettersOfMonthName = month.substring(0, 3).toUpperCase();
+        String regex = day + "\\s" + firstThreeLettersOfMonthName + "\\W*\\s\\d{4}";
```
Старайся все таки использовать String.format для таких штук. Более наглядно получается

alxiw replied:
Хорошо, везде поправил на него

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task03/CarsMailPage.java:
```diff
@@ -0,0 +1,68 @@
+
+    public CarsMailPage clickFirstPosAssessment() {
+        assessmentElement.click();
+        assessment = assessmentElement.findElement(By.xpath("//span[contains(@class, 'link__text margin_right_5')]")).getText();
```
Локаторы не нужно хранить в методе, так как на этапе поддержки тестов потом тяжело будет понять где и какой элемент юзается. Так же такой локатор совсем не пойдет, причину написал выше. Надо его переписать. И еще не стоит использовать классы типа margin_right_5, как пример, так как это автоматически генерируемые классы.

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task03/CarsMailPage.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    public CarsMailPage checkPopupTitle() {
+        assertTrue("Заголовок выпадающего окна не содержит названия параметра \"" + assessment + "\"",
```
String.format будет лучше

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task03/CarsMailPage.java:
```diff
@@ -0,0 +1,68 @@
+import static org.openqa.selenium.support.ui.ExpectedConditions.not;
+
+@UrlPattern("/catalog/ford/focus/.*")
+@PageUrl("/catalog/ford/focus/iii_restailing/sedan/")
```
машину передавать так хардкором не стоит. Так как завтра она может уже использоваться другая. Лучше передавай через параметры следующие параметры автомобиля: ford, focus, iii_restailing и sedan. Т.е. сюда будет передаваться что-то типа такого "/catalog/%1/%2/%3/%4"

alxiw replied:
Исправил

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task04/NewsSportMailPage.java:
```diff
@@ -0,0 +1,31 @@
+import static org.openqa.selenium.support.ui.ExpectedConditions.not;
+
+@UrlPattern("/news/football-worldcup/[\\d]{8}/")
+@PageUrl("/news/football-worldcup/32246552/")
```
Этот id надо передавать через параметры. Хранить надо в виде Enum. Я бы создал два перечисления. Один для типа игр футбола и там хранил football-worldcup в виде Football_worldcup("Чемпионат мира"), а вот во втором enum хранил бы CherkesovVybiraetSbornuyu("Так Черчесов будет выбирать состав сборной России на чемпионат мира", FootballType.Football_worldcup, 32246552). Таким образом ты связываешь параметры статьи отношением один ко многим.А в метод передавать enum с заголовком статьи и id

saint88 replied:
Не исправлено

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task05/ArticlePetsMailPage.java:
```diff
@@ -0,0 +1,33 @@
+import static org.junit.Assert.assertTrue;
+
+@UrlPattern("/how-to/.*")
+@PageUrl("/how-to/nyanya-dlya-sobak-i-koshek-kak-podgotovitsya-k-raz/")
```
Не нужно харкодить название статьи, а нужно передавать через параметры

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task05/ArticlePetsMailPage.java:
```diff
@@ -0,0 +1,33 @@
+@Domain("https://pets.mail.ru")
+public class ArticlePetsMailPage extends AbstractPage<ArticlePetsMailPage> {
+
+    @FindBy(xpath = "//time")
```
tagName="time"

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task01/PogodaMailPageTest.java:
```diff
@@ -0,0 +1,41 @@
+    public void add_city_in_favorites() {
+        new PogodaMailPage(driver)
+                .open()
+                .pageShallBeOpened()
```
Этот метод должен вызываться в опен

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -0,0 +1,39 @@
+                .open()
+                .pageShallBeOpened()
+                .openCalendar()
+                .checkMonthInCalendar("Июнь")
```
Через LocalDate

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -0,0 +1,39 @@
+    public void choose_date_in_calendar() {
+        new CalendarPogodaMailPage(driver)
+                .open()
+                .pageShallBeOpened()
```
Должен вызываться в open()

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -0,0 +1,39 @@
+                .openCalendar()
+                .checkMonthInCalendar("Июнь")
+                .goToPrevMonth()
+                .checkMonthInCalendar("Май")
```
Через LocalDate

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -0,0 +1,39 @@
+                .goToPrevMonth()
+                .checkMonthInCalendar("Май")
+                .goToNextMonth()
+                .checkMonthInCalendar("Июнь")
```
Через LocalDate

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -0,0 +1,39 @@
+                .checkMonthInCalendar("Июнь")
+                .goToPrevMonth()
+                .checkMonthInCalendar("Май")
+                .goToNextMonth()
```
В этот метод нужно передавать кнопку навигации

alxiw replied:
Написал выше

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -0,0 +1,39 @@
+                .pageShallBeOpened()
+                .openCalendar()
+                .checkMonthInCalendar("Июнь")
+                .goToPrevMonth()
```
Передавать надо в метод кнопку навигации

alxiw replied:
Написал выше

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task03/CarsMailPageTest.java:
```diff
@@ -0,0 +1,36 @@
+    public void open_and_close_cars_mail_popup() {
+        new CarsMailPage(driver)
+                .open()
+                .pageShallBeOpened()
```
вызываться должен в методе open()

alxiw replied:
Написал выше

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task04/NewsSportMailPageTest.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Test
+    public void check_error_on_news_sport_page() {
```
Я так понимаю это тест напроверку рекламных слотов, а где проверка этих слотов? Смотри если ты откроешь страничку, которая указана в кейсе и откроешь https://s.mail.ru/5bkc/nC4FCcDHC, то там в параметрах запроса передаются рекламные слоты. Так вот, чтобы их получить, тебе нужно включить сниффинг трафика на драйвере. Потом получить трафик и проверить, что у тебя в трафике есть вызовы тех рекламных слотов, которые указаны в кейсе. Должен быть метод, который принимает рекламные слоты и их параметры и проверяет их наличие. Но есть одно НО, драйвер складывает трафик в буфер и после получения его из этого буфера, он очищается. Поэтому тебе нужно получить весь трафик и потом его анализировать.

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task04/NewsSportMailPageTest.java:
```diff
@@ -0,0 +1,32 @@
+
+    @Test
+    public void check_error_on_news_sport_page() {
+        new NewsSportMailPage(driver)
```
Тест не соответствует кейсу и его надо переписать

saint88 replied:
@Step("Получаем трафик страницы")
    private List<Map<String, ?>> getPerformanceLogEntryByDomain(String domain, String method) {
        List<LogEntry> entries = getDriver().manage().logs().get(LogType.PERFORMANCE).filter(Level.ALL);

        return entries.parallelStream().filter(log -> log.getMessage().contains(domain) && log.getMessage().contains(method))
                .map(LogEntry::toJson).collect(Collectors.toList());
    }

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task04/SportMailPageTest.java:
```diff
@@ -0,0 +1,32 @@
+    }
+
+    @Test
+    public void check_error_on_sport_page(){
```
Этот тест не понятно что проверяет, но если рекламные слоты, то нужно сделать как написано выше и в одном тесте открывать сначала одну страничку, а потом другую

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task05/ArticlePetsMailPageTest.java:
```diff
@@ -0,0 +1,32 @@
+    public void check_existence_of_publication_date() {
+        new ArticlePetsMailPage(driver)
+                .open()
+                .pageShallBeOpened()
```
Должен вызываться в методе open()

alxiw replied:
Написал выше

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task05/ArticlePetsMailPageTest.java:
```diff
@@ -0,0 +1,32 @@
+        new ArticlePetsMailPage(driver)
+                .open()
+                .pageShallBeOpened()
+                .checkExistenceOfPublicationTime();
```
Вот тут неправильно так проверять. Ты проверяешь, что у тебя элемент виден на странице, но не проверяешь, что это дата (ее формат) и что текст в этом элементе не пустой, а ведь может же отображаться null например в элементе (поэтому надо проверить формат даты) или может вообще быть пусто, хотя сам элемент отображается на страничке (поэтому нужно проверять, что текст в элементе не exist). Хотя думаю если ты проверишь, только формат даты, то этого будет вполне достаточно.

alxiw replied:
Исправил

saint88 commented on homework-05/pom.xml:
```diff
@@ -0,0 +1,99 @@
+
+    <dependencies>
+        <dependency>
+            <groupId>ru.mail.track.ui11</groupId>
```
Сейчас так пойдет, но в реальной работе нужно выносить в отдельный мавен проект и зависимость делать на него, чтобы не тянуть кучу модулей, которые не имеют никакого отношения к данному модулю. Но вообще молодец =)

alxiw replied:
Вынесу тогда лучше эту основу из HW-04 в отдельный модуль и добавлю dependency на него в HW-04 и HW-05

saint88 reviewed (changes requested):
Нужно поправить замечания

alxiw committed:
Big project changes related to HW-04 and HW-05:
1. Create separate module based on common part used in both of the HWs
2. Implement most of changes requested in review
3. Update code style and improve readability

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task02/NewsPetsMailPage.java:
```diff
@@ -1,56 +1,61 @@
     @FindBy(css = "[class='pypo-item js-pgng_item']")
     private WebElement newsField;
 
+    private String newsFieldSelector = "[class='pypo-item js-pgng_item']";
```
Если поменяются классы местами, то этот селектор работать не будет. Надо его поменять

alxiw replied:
Правильно понимаю, что если нет у элементов уникальных имён или id, то лучше всего, например, касаемо данного теста, найти заголовок "Новости", подняться на пару-тройку уровней выше, перейти к соседнему и спуститься к item'ам новостей?

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task02/NewsPetsMailPage.java:
```diff
@@ -1,56 +1,61 @@
     private WebElement newsField;
 
+    private String newsFieldSelector = "[class='pypo-item js-pgng_item']";
+    private String showMoreButtonSelector = "[class='button js-pgng_more_link']";
```
Аналогично выше

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
     @FindBy(css = "[class='suggest__inner js-suggest__render'")
     private WebElement suggestMenu;
 
+    private String suggestNamesLocator = "//div[@class='newsitem newsitem_vertical newsitem_special newsitem_border_bottom']";
```
Если поеняют классы местами, то этот xpath работать не будет. Надо его поменять

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
     private WebElement suggestMenu;
 
+    private String suggestNamesLocator = "//div[@class='newsitem newsitem_vertical newsitem_special newsitem_border_bottom']";
+    private String nameInfoElement = "//div[@class='newsitem newsitem_vertical newsitem_namesecret']";
```
Аналогично выше

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
 
+    private String suggestNamesLocator = "//div[@class='newsitem newsitem_vertical newsitem_special newsitem_border_bottom']";
+    private String nameInfoElement = "//div[@class='newsitem newsitem_vertical newsitem_namesecret']";
+    private String letterOpenButtonSelector = "[class='filter__item filter__item_active']";
```
аналогично выше

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
+    private String suggestNamesLocator = "//div[@class='newsitem newsitem_vertical newsitem_special newsitem_border_bottom']";
+    private String nameInfoElement = "//div[@class='newsitem newsitem_vertical newsitem_namesecret']";
+    private String letterOpenButtonSelector = "[class='filter__item filter__item_active']";
+    private String suggestElementLocator = "//div[@class ='suggest__item js-suggest__item']";
```
аналогично выше

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
+    private String nameInfoElement = "//div[@class='newsitem newsitem_vertical newsitem_namesecret']";
+    private String letterOpenButtonSelector = "[class='filter__item filter__item_active']";
+    private String suggestElementLocator = "//div[@class ='suggest__item js-suggest__item']";
+    private String genderSelectLocator = "//div[@class ='dropdown dropdown_scrollable dropdown_scrollable dropdown_scrollable js-select js-module']";
```
Аналогично выше

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
+    private String letterOpenButtonSelector = "[class='filter__item filter__item_active']";
+    private String suggestElementLocator = "//div[@class ='suggest__item js-suggest__item']";
+    private String genderSelectLocator = "//div[@class ='dropdown dropdown_scrollable dropdown_scrollable dropdown_scrollable js-select js-module']";
+    private String genderListLocator = "//div[@class ='suggest__inner js-select__options__list js-scrollable__view dropdown__scroll']";
```
Аналогично выше. Не нужно привязываться к следованию значений атрибутов элемента. Поищи уникальные классы или комбинации. Они могут находиться так же в элементах ниже или выше. Надо переписать этот локатор

saint88 commented on homework-04/src/main/java/ru/mail/track/ui11/hw04/task03/NameSecretHoroMailPage.java:
```diff
@@ -29,87 +34,88 @@
-        String womanSelectElementLocator = "//div[@class ='suggest__item js-select__options__item' and text()='Женские']";
+    public NameSecretHoroMailPage selectGenderInMenu(String gender) {
+        logger.log("Select gender in menu");
+        String womanSelectElementLocator = String.format("//div[@class ='suggest__item js-select__options__item' and text()='%s']", gender);
```
Это плохой локатор, аналогично как написано выше. Так же локаторы лучше выносить в свойства объекта страницы и инкапсулировать (PageObject) и не нужно хранить их в методах, так как они могут понадобиться в других методах или например могут быть частью других локаторов. Да и поддерживать такие локаторы намного проще и легче.

saint88 commented on homework-04/src/test/java/ru/mail/track/ui11/hw04/task04/ArticlePageTest.java:
```diff
@@ -26,7 +25,8 @@ public void killBrowser() {
         new ArticlePage(driver)
-                .open("articles", "economic")
-                .checkPageUrl();
+                .open("articles", "some")
```
А что этот тест проверяет? В задании же написано, что нужно проверять наличие рекламных слотов в параметрах запроса. Я даже тебе скриншот прислал. Посмотри в лекциях тему как включить сниффер трафика на драйвере. Так же получи запросы (тоже писал в слаке) и затем проверь, что когда открывается страничка, то у тебя есть запрос с теми параметрами (номер слота, site_id и т.д.), которые указаны в кейсе из дз.

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -64,18 +66,19 @@ public PogodaMailPage typeSearchText(String text) {
 
-    public PogodaMailPage pressCity() {
+    public PogodaMailPage pressFirstSuggestedCity() {
+        String firstSuggestedCity = String.format("//span[@class='pm-toolbar__suggests__group__item  pm-toolbar__suggests__group__item_with-caption']//span[text()='%s']", city);
```
Во первых локатор такой не пойдет, написал причину выше. Во вторых не нужно хранить локаторы в методе, а нужно хранить их в свойствах объекта страницы (паттерн PageObject)

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task01/PogodaMailPage.java:
```diff
@@ -85,7 +88,7 @@ public PogodaMailPage pressCity() {
         assertTrue("Город отображается на странице",
                 standardWaiter.waitForCondition((
-                        ExpectedConditions.visibilityOf(cityOnPage))));
+                        ExpectedConditions.visibilityOf(driver.findElement(By.xpath(String.format("//div[@class='information__header__left__place']/h1[text()='%s']", city)))))));
```
все локаторы должны быть свойствами объекта страницы. И еще проверь локатор //h1, что-то мне подсказывает, что на странице всего лишь один заголовок первого уровня и твой локатор очень сложный и перегруженный

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPage.java:
```diff
@@ -41,36 +38,31 @@ public CalendarPogodaMailPage openCalendar() {
 
     public CalendarPogodaMailPage goToDay(int day) {
-        driver.findElement(By.xpath("//table[contains(@class, 'calendar')]//a[text() = '" + day + "']")).click();
+        driver.findElement(By.xpath(String.format("//table[contains(@class, 'calendar')]//a[text() = '%d']", day))).click();
```
не нужно хранить локаторы в методе, лучше их сделать в свойствах объекта страницы

saint88 commented on homework-05/src/main/java/ru/mail/track/ui11/hw05/task03/CarsMailPage.java:
```diff
@@ -5,20 +5,20 @@
 public class CarsMailPage extends AbstractPage<CarsMailPage> {
 
-    @FindBy(xpath = "//a[@data-type-tag = 'plus']")
+    @FindBy(xpath = "//span[contains(text(), 'Плюсы')]/following::span[contains(text(), 'Управляемость')][1]")
```
А для минусов нельзя использовать этот же локатор, только с другими данными в шаблоне вместо "Плюсы" и "Управляемость"? А то получается у нас для каждого параметра мы будем писать одинаковые локаторы. В общем локатор должен быть динамический и параметры надо передавать через enum

saint88 commented on homework-05/src/test/java/ru/mail/track/ui11/hw05/task02/CalendarPogodaMailPageTest.java:
```diff
@@ -24,16 +30,22 @@ public void killSession() {
         new CalendarPogodaMailPage(driver)
-                .open()
-                .pageShallBeOpened()
+                .open("prognoz", "easter_island",
```
в тесте не стоит делать таку логику. Лучше ее сделать на пейдже. И там уже делай эту логику. А в самом тесте должен быть просто вызов метода open("easter_island", first)

saint88 reviewed (changes requested):
Надо поправить локаторы. Не нужно привязываться к порядку следования классов в элементе. И до поправлять остальные замечания.

alxiw committed:
Update following to Page Object pattern and selectors in homeworks

alxiw commented:
Исправил селекторы в четвёртом и пятом заданиях, а также улучшил следование паттерну Page Object
Проверку логов обращения по сети решил пока отложить, поскольку приоритетнее сейчас остальные ДЗ с проектом, но если до экзамена время останется, то обязательно вернусь к этому вопросу

saint88 reviewed (approved):
Все отлично. Ставлю 8 баллов, так как нет задачи с логами. Сделаешь ее, если время останется, исправлю на 10. Но 8 баллов это тоже отлично.
